api/wasm/cpp/proxy_wasm_intrinsics.h

```diff
@@ -247,6 +312,85 @@ struct Tuple3Hash {
   }
 };
 
+using HeaderStringPairs = std::vector<std::pair<std::string, std::string>>;
+
+class GrpcCallHandlerBase {
+public:
+  GrpcCallHandlerBase(Context* context) : context_(context) {}
+  virtual ~GrpcCallHandlerBase() {}
+
+  void cancel();
+
+  virtual void onCreateInitialMetadata() = 0;
+  virtual void onSuccess(std::unique_ptr<WasmData> message) = 0;
+  virtual void onFailure(GrpcStatus status, std::unique_ptr<WasmData> error_message) = 0;
+
+private:
+  friend class Context;
+
+  Context* const context_;
+  uint32_t token_;
+};
+
+template<typename Message>
+class GrpcCallHandler : public GrpcCallHandlerBase {
+public:
+  GrpcCallHandler(Context* context) : GrpcCallHandlerBase(context) {}
+  virtual ~GrpcCallHandler() {}
+
+  virtual void onSuccess(Message&& response) = 0;
+
+private:
+  void onSuccess(std::unique_ptr<WasmData> message) override {
+    onSuccess(message->proto<Message>());
+  }
+};
+
+class GrpcStreamHandlerBase {
+public:
+  GrpcStreamHandlerBase(Context* context) : context_(context) {}
+  virtual ~GrpcStreamHandlerBase() {}
+
+  // NB: with end_of_stream == true, callbacks can still occur: reset() to prevent further callbacks.
+  void send(std::string_view message, bool end_of_stream);
+  void close(); // NB: callbacks can still occur: reset() to prevent further callbacks.
```

Comment:
Same as above, shouldn't we automatically remove callbacks on remote close, then?

Reply:
Can't unfortunately because "close()" here is really just HalfClose in TCP terminology which means that it is only informing the other side of the close of writing on this side event, so it doesn't prevent the other side from sending data until that side decides to close.  Yes, it is pretty lame that they chose to use a different name for an existing operation, but there we are.

Reply:
Got it, thanks.